model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsWorker.java

```diff
@@ -42,83 +38,85 @@ public class PersistentSessionsWorker {
-            Collection<PersistentDeferredElement<K, V>> batch = new ArrayList<>();
-            PersistentDeferredElement<K, V> polled = queue.poll(100, TimeUnit.MILLISECONDS);
+        private void process(ArrayBlockingQueue<PersistentUpdate> queue) throws InterruptedException {
+            ArrayList<PersistentUpdate> batch = new ArrayList<>();
+            PersistentUpdate polled = queue.poll(100, TimeUnit.MILLISECONDS);
             if (polled != null) {
                 batch.add(polled);
-                queue.drainTo(batch, 99);
+                queue.drainTo(batch, maxBatchSize - 1);
                 try {
                     LOG.debugf("Processing %d deferred session updates.", batch.size());
-                    KeycloakModelUtils.runJobInTransaction(factory,
-                            session -> adapter.run(((PersistentUserSessionProvider) session.getProvider(UserSessionProvider.class)), batch, offline));
+                    Retry.executeWithBackoff(iteration -> {
+                                if (iteration < 2) {
+                                    // attempt to write whole batch in the first two attempts
+                                    KeycloakModelUtils.runJobInTransaction(factory,
+                                            innerSession -> batch.forEach(c -> c.perform(innerSession)));
+                                    batch.forEach(PersistentUpdate::complete);
+                                } else {
+                                    LOG.warnf("Running single changes in iteration %d for %d entries", iteration, batch.size());
+                                    ArrayList<PersistentUpdate> performedChanges = new ArrayList<>();
+                                    List<Throwable> throwables = new ArrayList<>();
+                                    batch.forEach(change -> {
+                                        try {
+                                            KeycloakModelUtils.runJobInTransaction(factory,
+                                                    change::perform);
+                                            change.complete();
+                                            performedChanges.add(change);
+                                        } catch (Throwable ex) {
+                                            throwables.add(ex);
+                                        }
+                                    });
+                                    batch.removeAll(performedChanges);
+                                    if (!throwables.isEmpty()) {
+                                        RuntimeException ex = new RuntimeException("unable to complete some changes");
+                                        throwables.forEach(ex::addSuppressed);
+                                        throw ex;
+                                    }
+                                }
+                            },
+                            Duration.of(10, ChronoUnit.SECONDS), 0);
```

Comment:
0 means we add no backoff, is it correct? If there are more nodes in two DCs fighting to update the same entity 0 can make it longer. Maybe we can keep this 0 by default but make it configurable so we can test it within A/A perf tests?

Reply:
In the recent tests when every update was going through the background thread, I saw no deadlocks any more. I would assume that if two nodes fight, one of them would win and it would be ok to retry immediately, so I chose no back-off period. Maybe we would just keep an eye on it and make it configurable when we see the need in one of our performance tests?